extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcUtils.java

```diff
@@ -0,0 +1,80 @@
+package io.quarkus.oidc.runtime;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import io.quarkus.oidc.OIDCException;
+import io.quarkus.oidc.OidcConfig;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+public final class OidcUtils {
+    private OidcUtils() {
+
+    }
+
+    public static List<String> findRoles(String clientId, OidcConfig.Roles rolesConfig, JsonObject json) throws Exception {
+        // If the user configured a specific path - check and enforce a claim at this path exists
+        if (rolesConfig.getRoleClaimPath().isPresent()) {
+            return findClaimWithRoles(rolesConfig, rolesConfig.getRoleClaimPath().get(), json, true);
+        }
+
+        // Check 'groups' next
+        List<String> groups = findClaimWithRoles(rolesConfig, "groups", json, false);
+        if (!groups.isEmpty()) {
+            return groups.stream().map(v -> v.toString()).collect(Collectors.toList());
+        } else {
+            // Finally, check if this token has been issued by Keycloak.
+            // Return an empty or populated list of realm and resource access roles
+            List<String> allRoles = new LinkedList<>();
+            allRoles.addAll(findClaimWithRoles(rolesConfig, "realm_access/roles", json, false));
+
+            for (String resourceRole : findClaimWithRoles(rolesConfig, "resource_access/frontend/roles", json, false)) {
```

Comment:
The `frontend` in `resource_access/frontend/roles` is the client id. So you would need to process `resource_access` path where the first child is the client and then the roles.

Reply:
@pedroigor I don't think I follow :-).
So the case 1 is what was there originally, `realm_access/roles/`.

now there is another case with `resource_access`. Can you please expand on it with an example showing how both `realm_access/roles/` and `resource_access/{frontend}/roles/` cooperate ?